src/Message/AbstractResponse.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+
+namespace Omnipay\BlueSnap\Message;
+
+use Omnipay\BlueSnap\Transaction;
+use Omnipay\Common\Message\RequestInterface;
+use SimpleXMLElement;
+
+abstract class AbstractResponse extends \Omnipay\Common\Message\AbstractResponse
+{
+    /**
+     * HTTP request id
+     *
+     * @var string|null
+     */
+    protected $requestId;
+
+    /**
+     * HTTP response code
+     *
+     * @var string|null
+     */
+    protected $code;
+
+    /**
+     * @var string|SimpleXMLElement|array
+     */
+    protected $data;
+
+    /**
+     * @var Transaction|null
+     */
+    protected $transaction;
+
+    /**
+     * @param RequestInterface $request the initiating request.
+     * @param mixed $data
+     */
+    public function __construct(RequestInterface $request, $data)
+    {
+        parent::__construct($request, $data);
+        $this->transaction = null;
+    }
+
+    /**
+     * Returns true if the request was successful, false otherwise
+     *
+     * @return bool
+     */
+    public function isSuccessful()
+    {
+        return substr($this->getCode() ?: '', 0, 1) === '2';
+    }
+
+    /**
+     * Get the ID from the HTTP request
+     *
+     * @return string|null
+     */
+    public function getRequestId()
+    {
+        return $this->requestId;
+    }
+
+    /**
+     * Set the ID from the HTTP request
+     *
+     * @param string|null $requestId
+     * @return static
+     */
+    public function setRequestId($requestId)
+    {
+        $this->requestId = $requestId;
+        return $this;
+    }
+
+    /**
+     * Get the HTTP response code
+     *
+     * @return string|null
+     */
+    public function getCode()
+    {
+        return $this->code;
+    }
+
+    /**
+     * Set the HTTP response code
+     *
+     * @param string $code
+     * @return static
+     */
+    public function setCode($code)
+    {
+        $this->code = $code;
+        return $this;
+    }
+
+    /**
+     * Overriding to provide more specific return type
+     *
+     * @return \Omnipay\BlueSnap\Message\AbstractRequest
+     */
+    public function getRequest()
+    {
+        /**
+         * @var \Omnipay\BlueSnap\Message\AbstractRequest
+         */
+        return parent::getRequest();
+    }
+
+    /**
+     * Get the error message from the response. Returns null if request was successful.
+     *
+     * Since this library works with both XML and JSON response data, this function checks for both types.
+     *
+     * @return string|null
+     * @psalm-suppress MixedPropertyFetch because we check the data typing before using.
+     */
+    public function getMessage()
```

Comment:
True. There will be some redundancy if I force every response to implement this but I could split it into a `GetXML` and `GetJSON` traits and have it be shared among response classes

Reply:
Yeah I like that approach! I don't think logic like this should be handled in an abstract class aside from the abstract method being defined, or it could be omitted here and included via a trait